### ISSUE TITLE
Make the effects panel accessible

### ISSUE THREAD

**Issue by LWinterberg** [audacity/audacity#3078]
# Shortcut

Users should be able to hide / show the ‘Realtime effects’ panel by pressing the ‘e’ key (without modifier). Screenreader reads: “Realtime effects for `track name`”

Pressing `e` again or `ESC` closes the panel and re-focuses on the track. 

# Keyboard Navigation

Tab cycles through the elements (black outline), in the order:
1. Panel header (reads title) 
2. Panel power button
3. Close effect panel 
4. Effects slot 1 (whole slot)
5. Effect Power Button
6. Effect Dropdown
7. Effect Settings
8. Effects slot 2

...
n-1. Add New Effect
n. (the text info about realtime effects)
n+1. cycles back to the top (Panel header)

Further, Up/Down can be used to navigate between effect slots. Right is equivalent to Down, Left is equivalent to Up. 

<details><summary>examples</summary>

![image](https://user-images.githubusercontent.com/87814144/175000497-08e4c0f0-51b4-47ee-93da-6f85331f71a5.png)
![image](https://user-images.githubusercontent.com/87814144/175000527-f0fbe59c-e20c-436d-887c-ce2137776585.png)
![image](https://user-images.githubusercontent.com/87814144/175000553-71191427-c5fe-4a41-bba1-29e9fe51fbdc.png)

</details>

# Reordering Effects

When an effect slot is selected, pressing Alt+Down/Right moves it down, Alt+Left/Up moves it up. 

# Selecting Items
* The bypass can be activated/deactivated with Enter. 
* The dropdown can be opened with Enter
* The settings can be opened with Enter. 

# Outlines
The normal wx focus outline is used. 
<details><summary>examples</summary>

![image](https://user-images.githubusercontent.com/87814144/173855660-26ba77ec-69cb-49d5-8079-38588aa11e24.png)
![image](https://user-images.githubusercontent.com/87814144/173855699-b27eb445-aaf3-4fe5-a777-66c5a6121069.png)
![image](https://user-images.githubusercontent.com/87814144/173855757-75dfe340-1e12-4d7b-888f-b07eeb93eb8d.png)

</details>

**Comment by DavidBailes** [comment]
I have previously suggested that it would be better if the list of effects in the effects panel had a keyboard interaction appropriate for a list, and was read by screen readers as a list. I appreciate that wxWidgets does not supply any list classes where each list item contains a number of controls, and so a custom list would have to be constructed.
Is the decision that the list does not interact as a list for keyboard users and users of screen readers due to a disagreement of this in principle, or shortage of developer time or what?

**Comment by LWinterberg** [comment]
A mix of all of the above: I tried the list navigation they had over at musescore 4, and found that, while faster to move around, it also becomes a lot more difficult to understand what exactly it is you need to do while moving around. Like, I can tell I'm navigating a list, but it's unclear how to access the actions (or that this list even has actions). So in the interest of not having to develop and test entirely new patterns for this already over-full release, I decided to go with the tab version for the time being.

**Comment by Ryan-Miller-Audacity** [comment]
I'm not 100% about accessibility guidelines for Desktop applications, but should the new "watch video" link within the panel need a image to denote that this link opens in a new window (example shown below)? In a similar vein, we should be sure that screenreaders properly announce that the link opens in a new window as well correct? 

This is the type of symbol that I have used/seen in previous work, although this was all web-based accessibility standards and not specifically for desktop applications.

![link opens in new window](https://user-images.githubusercontent.com/88568755/174147801-03d77f29-0b02-49c4-9aed-03afb8171ee6.png)

**Comment by DavidBailes** [comment]
> A mix of all of the above: I tried the list navigation they had over at musescore 4, and found that, while faster to move around, it also becomes a lot more difficult to understand what exactly it is you need to do while moving around. Like, I can tell I'm navigating a list, but it's unclear how to access the actions (or that this list even has actions).

Which list, or lists, in musescore 4 did you find confusing to navigate?

Here's an outline of my suggestion for the keyboard navigation of the list of effects in the effects pane, where each list item contains three buttons:
1. If a list item (the whole list item) is the focus, then you can move to other list items by using the standard keystrokes for lists: up/down arrow, home, end.
2. If the focus is a list item and you press tab, this moves you to the first of the buttons. Continuing to press tab takes you to the second and third buttons, and then the control after the list, which in this case is the add effects button.
3. If one of the buttons in a list item is the focus, then pressing up or down arrow moves you to the previous or next list item (the whole list item), and similarly, pressing home or end moves you to the first or last list item.
4. A list item has a context menu, which includes the commands for moving the effect up or down. The menu could also include other commands which duplicate some of the functions of the buttons if this was considered that it would be convenient to the user. For example to remove the effect from the list. (the menu button opens a menu containing commands which would often be found on an items context menu).
5. The list would have a custom accessible object, so that the list is read to screen readers as being a list.

Concerning your comment about navigation in a musescore 4 list " but it's unclear how to access the actions (or that this list even has actions)", in the above scheme, a user automatically discoverers the buttons  in a list item when they tab expecting to go to the control after the list.

Examples of the above scheme (or very close) are:
- the list of installed extensions in visual studio code
- the list of notifications in the action centre in Windows 10.

**Comment by DavidBailes** [comment]
> I'm not 100% about accessibility guidelines for Desktop applications, but should the new "watch video" link within the panel need a image to denote that this link opens in a new window (example shown below)? In a similar vein, we should be sure that screenreaders properly announce that the link opens in a new window as well correct?

My guess is that most users would expect a "watch video" link to open a web browser to show the video, and that in this case this doesn't have to be explicitly stated. I think it will be of far more interest to users of screen readers as to whether the video describes how to use the effects panel using the keyboard as well as using the mouse.

**Comment by LWinterberg** [comment]
Hm, I may have misunderstood what you meant the first time around. I'll hand this decision over to @Tantacrul as he's more experienced in these things than i am.

**Comment by Tantacrul** [comment]
Nice point @DavidBailes. I'll add keyboard control to my script.

**Comment by Tantacrul** [comment]
@DavidBailes - a few comments:

> If a list item (the whole list item) is the focus, then you can move to other list items by using the standard keystrokes for lists: up/down arrow, home, end.
I think we can make it function like a list item although I have one or two questions about exact functionality.

> If the focus is a list item and you press tab, this moves you to the first of the buttons. Continuing to press tab takes you to the second and third buttons, and then the control after the list, which in this case is the add effects button.

Would it not be more generally consistent to use **Space & Enter** to 'select' the effect and therefore dig deeper into it? Then up/down would navigate the options. Esc. would then bring the user back to the general list of effects (which they could continue to navigate through using 'down/up'

> A list item has a context menu, which includes the commands for moving the effect up or down. The menu could also include other commands which duplicate some of the functions of the buttons if this was considered that it would be convenient to the user. For example to remove the effect from the list. (the menu button opens a menu containing commands which would often be found on an items context menu).

I'd be inclined to just provide shortcuts for moving effects up down and deleting (**delete & backspace** to remove, **ctrl+up/down** (or similar) to re-stack? I'm not against a context menu either, to be honest.

> The list would have a custom accessible object, so that the list is read to screen readers as being a list.
Agreed - I guess the whole effects panel, when first triggered, would be in focus and would then list out its contents

**Comment by DavidBailes** [comment]
> @DavidBailes - a few comments:
> 
> > If a list item (the whole list item) is the focus, then you can move to other list items by using the standard keystrokes for lists: up/down arrow, home, end.
> 
> I think we can make it _function_ like a list item without it actually being one.

Sure, it would have to be a custom list with the desired behaviour.

 I have one or two questions & thoughts about the exact functionality. Let me know what you think.
> 
> > If the focus is a list item and you press tab, this moves you to the first of the buttons. Continuing to press tab takes you to the second and third buttons, and then the control after the list, which in this case is the add effects button.
> 
> Would it not be more generally consistent to use **Space & Enter** to 'select' the effect and therefore dig deeper into it? Then up/down would navigate the options. **esc** would then bring the user back to the general list of effects (which they could continue to navigate using 'down/up')

Maybe on Macs but as far as I'm aware I don't think it would be consistent with anything on Windows. Can you give me an example of your suggested behaviour on Windows?
The only thing that springs to mind is in the Windows settings app, time & language > language, there's a list of preferred languages. Pressing Enter on an item reveals controls for that language. But you tab to those controls and you can't "escape" back out of those controls.
I gave a couple of examples where list items contained controls on Windows, and they don't follow your suggested behaviour.

> 
> > A list item has a context menu, which includes the commands for moving the effect up or down. The menu could also include other commands which duplicate some of the functions of the buttons if this was considered that it would be convenient to the user. For example to remove the effect from the list. (the menu button opens a menu containing commands which would often be found on an items context menu).
> 
> I'd be inclined to just provide shortcuts for moving effects up/down and deleting them. Namely: **delete & backspace** to remove, **ctrl+up/down** (or similar) to re-stack? I'm not against a context menu either, to be honest.

I don't think there are any standard shortcuts for moving items up and down, so letting the user know what they are for this list by displaying them after the commands in a context menu would be helpful, I think.

> 
> > The list would have a custom accessible object, so that the list is read to screen readers as being a list.
> 
> Agreed - I guess the whole effects panel, when first triggered, would be in focus and would then list out its contents

I don't think that would cause it's contents to be read out, but you never know with screen readers. If there were no effects in the list, then it might be convenient to make the add effect button the initial focus. (The accessibility name of the button could include the name of the track.)
